cfpb/cfgov-refresh
# Mega Menu in Wagtail

Uses Wagtail models to manage mega menu content instead of `_vars-maga-menu.html` (more information in GHE issue 1120.)

## Additions
- `MenuItem` model
- `Link`, `NavGroup`, and `NavItem` blocks


## Changes
- Updates to mega menu template

## Testing

### New instructions
1. Check out branch and run `gulp` & `python cfgov/manage.py migrate`
2. Verify the recent changes to the menu in commit https://github.com/cfpb/cfgov-refresh/pull/3396/commits/ec78bc4fe2c7d76ac0a895c2a9fe56dc654f04d0 and at localhost:8000 (the hard-coded `vars-mega-menu` version of the menu will be active)
3. Set the `WAGTAIL_MENU` flag to activate the wagtail version of the menu
4. Verify that no menu items exist at localhost:8000
5. Run `python cfgov/manage.py runscript migrate_mega_menu` to generate the existing menu in Wagtail
6. Verify the Wagtail version of the menu at localhost:8000
7. Try out the `Mega Menu` admin interface. Due to its permissions settings, you will need to be logged in as a superuser to view/edit the `Mega Menu` section of the Wagtail admin (run `python cfgov/manage.py createsuperuser` to create a superuser if necessary)

### Old instructions
1. Check out branch
2. Run `python cfgov/manage.py migrate`
3. Run `gulp`
4. Switch to Wagtail-based mega menu by setting `WAGTAIL_MENU` flag to True
5. Test menu item snippets (note: due to its permissions settings, you will need to be logged in as a superuser to view the Mega Menu section of the Wagtail admin)
    - Navigate to [Mega menu item admin](http://localhost:8000/admin/v1/menuitem/) and add a `Menu Item` with some content. 
    - Navigate to [a Wagtail page](http://localhost:8000/) and verify that the content displays in the menu.
6. Test draft state
    - Create a menu item, add a `Nav Group` block in Column 1, and check its `Draft` checkbox. Add links with `Live` and `Draft` state.
    - Add another `Nav Group` block in Column 1 and don't check the Draft option. Add links with `Live` and `Draft` state.
    - Navigate to [a Wagtail page](http://localhost:8000/) and verify that the`Nav Group` block and any links with Draft state are not displayed.
    - Set up local sharing site 
    - Verify that `Nav Group` block and links with draft state now show instead of the ones with Live state.
7. To populate Wagtail-based menu with existing content, run `python cfgov/manage.py runscript migrate_mega_menu`

## Screenshots

### Snippets with menu item option
<img width="567" alt="8df99e80-a1a1-11e7-8dc2-fa5c36b1f698" src="https://user-images.githubusercontent.com/778171/31030821-89bfdaba-a50b-11e7-8225-e978e5bbd834.png">

### Menu item snippet form
Provides inputs for the menu item's link text, URL, and order within the menu and streamfields for an optional footer and up to 4 columns of menu content.
![ac151d40-a1a1-11e7-945b-b424e8e8a6b7](https://user-images.githubusercontent.com/778171/31031172-cf6699d6-a50c-11e7-9520-b32cece2868d.png)

### Column/Footer streamfield blocks
Multiple blocks can be added, but only the last valid one in each streamfield will be displayed. 

#### Nav Group block 
Available in each of the four column streamfields
<img width="1178" alt="screen shot 2017-09-29 at 11 46 38 am" src="https://user-images.githubusercontent.com/778171/31031205-e1ef7898-a50c-11e7-85b2-387b7a1ab9d3.png">

#### Featured Content Module block
Available in 4th column as alternative to links
<img width="1063" alt="cb2f7018-a1a1-11e7-971a-745c520e182c" src="https://user-images.githubusercontent.com/778171/31031604-e3e529c6-a50d-11e7-91b1-f3ce2b89ed26.png">

#### Footer block
Optional content that displays below the nav items
<img width="1062" alt="d2cc0b1a-a1a1-11e7-99bc-12831b0e55c3" src="https://user-images.githubusercontent.com/778171/31031607-e60b9f32-a50d-11e7-9b37-aae11514e423.png">

### Draft state
Since snippets don't have draft state the way Wagtail pages do, this PR explores a couple of options for adding draft functionality using a combination of feature flags and state inputs.

#### Draft checkbox on Nav Group/FCM/Footer blocks
When this box is checked, the block will only show if the `DRAFT_MENU` flag is on.
<img width="833" alt="screen shot 2017-09-29 at 12 21 16 pm" src="https://user-images.githubusercontent.com/778171/31032420-c19ce0c2-a510-11e7-9d85-319395564055.png">

#### Live/Draft/Both state on links
`Draft` links show when the `DRAFT_MENU` flag is on; `Live` links show when the flag is off.
<img width="782" alt="screen shot 2017-09-29 at 12 21 25 pm" src="https://user-images.githubusercontent.com/778171/31032430-c3ed1f90-a510-11e7-8fc3-3af6e6632840.png">




## Todos

- Tests

## Checklist
